### PR TITLE
OSM (hardcoded) + REST Sail 

### PR DESCRIPTION
**Pull request by gspinaci**
This pull request contains the java code to execute and run OpenStreetMap federation. It comes with its own documentation page.  The OSM federation integration is completely encoded within the source files.

It contains also a starting RESTSail federation integration, which is a general federation to handle any type of REST request. Its documentation is still under development.

Right now, RESTSail only handles GET and POST methods and it is able to parse only JSON responses.

### CODE REVIEW

**Comment by aindlq** [comment]
Nice! :+1: 

I tested OSM sail and it works perfectly:
image

Actually, I would like to have the OSM repository enabled by default, but to do this we need to make sure that we follow Nominatim Usage Policy - https://operations.osmfoundation.org/policies/nominatim/.
So we need to find a way to add rate limit and propagate HTTP Referer and User-Agent.

**Comment by aindlq** [comment]
@gspinaci fyi, I've found how to do rate limiting, so it in on me.

**Comment by gspinaci** [comment]
I have generalized the parser behavior to handle JSON objects

File: src/main/java/org/researchspace/sail/rest/AbstractRESTWrappingSailConnection.java [hunk 1]
```diff
@@ -62,9 +66,8 @@ public AbstractRESTWrappingSailConnection(AbstractServiceWrappingSail sailBase)
      */
     protected Response submit(RESTParametersHolder parametersHolder) {
         try {
-            Client client = ClientBuilder.newClient();
-            WebTarget targetResource = client.target(getSail().getUrl()).property(ClientProperties.FOLLOW_REDIRECTS,
-                    Boolean.TRUE);
+            WebTarget targetResource = this.getSail().getClient().target(getSail().getConfig().getUrl())
+                    .property(ClientProperties.FOLLOW_REDIRECTS, Boolean.TRUE);
```
Comment (aindlq) [inline src/main/java/org/researchspace/sail/rest/AbstractRESTWrappingSailConnection.java:70]: just an observation fyi, it is not going to follow redirect from http->https, and will just fail. Which, I believe is OK. See  https://stackoverflow.com/questions/21778393/jersey-jax-rs-client-2-5-follow-redirect-from-http-to-https and https://stackoverflow.com/questions/1884230/httpurlconnection-doesnt-follow-redirect-from-http-to-https

File: src/main/java/org/researchspace/sail/rest/RESTSailConnection.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        List<BindingSet> bindingSets = Lists.newArrayList();
+
+        // Iterate over each object in JSONArray and evaluate its
+        for (Object object : array) {
+            MapBindingSet mapBindingSet = new MapBindingSet();
+
+            // Evaluate each jsonPath singularly and add the result to the binding
+            for (Map.Entry<String, String> path : jsonPaths.entrySet()) {
+                mapBindingSet.addBinding(path.getKey(),
+                        VF.createLiteral(JsonPath.read(object, path.getValue()).toString(), XSD.STRING));
```
Comment (aindlq) [inline src/main/java/org/researchspace/sail/rest/RESTSailConnection.java:145]: we shouldn't always use XSD.STRING, maybe as a default is OK, but if defined we should take the value from `spl:valueType`, e.g as in:
```
    spin:column [
        a spin:Column ;
        rdfs:comment "Display Name" ;
        spl:predicate :_display_name ;
        spl:valueType xsd:string ;
        ephedra:jsonPath "$.display_name"
        ] ;
```

File: src/main/java/org/researchspace/sail/rest/RESTSailConnection.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+     * @return
+     */
+    private List<BindingSet> iterateJsonMap (Map map, Map<String, String> jsonPaths) {
+       
+        List<BindingSet> bindingSets = Lists.newArrayList(); 
+        MapBindingSet mapBindingSet = new MapBindingSet();
+        
+        for(Map.Entry<String, String> path : jsonPaths.entrySet()) {
+            Object object = JsonPath.read(map, path.getValue());
+            mapBindingSet.addBinding(path.getKey(), VF.createLiteral(object.toString(), XSD.STRING));
```
Comment (aindlq) [inline src/main/java/org/researchspace/sail/rest/RESTSailConnection.java:169]: same as above

**Review by aindlq — APPROVED** [review]
looks good, I think we can address all comments in the followup

File: src/main/java/org/researchspace/sail/rest/RESTSailConnection.java (hunk 2 above)
Comment (gspinaci) [reply src/main/java/org/researchspace/sail/rest/RESTSailConnection.java:145]: See https://github.com/researchspace/researchspace/commit/89b5639bc36a559d8c701105a087ee8e321ead70

File: src/main/java/org/researchspace/sail/rest/RESTSailConnection.java (hunk 3 above)
Comment (gspinaci) [reply src/main/java/org/researchspace/sail/rest/RESTSailConnection.java:169]: See https://github.com/researchspace/researchspace/commit/89b5639bc36a559d8c701105a087ee8e321ead70

**Pull request merged**